pkg/pubsub/redis/stream_puller.go

```diff
@@ -0,0 +1,563 @@
+	}
+}
+
+// WithStreamPullerBatchSize sets the batch size for reading from the stream
+func WithStreamPullerBatchSize(size int64) StreamPullerOption {
+	return func(p *StreamPuller) {
+		p.batchSize = size
+	}
+}
+
+// WithStreamPullerBlockTime sets the block time for reading from the stream
+func WithStreamPullerBlockTime(d time.Duration) StreamPullerOption {
+	return func(p *StreamPuller) {
+		p.blockTime = d
+	}
+}
+
+// WithStreamPullerIdleTime sets the idle time for detecting stale messages
+func WithStreamPullerIdleTime(d time.Duration) StreamPullerOption {
+	return func(p *StreamPuller) {
+		p.idleTime = d
+	}
+}
+
+// WithStreamPullerPartitionCount sets the number of partitions for the puller
+func WithStreamPullerPartitionCount(count int) StreamPullerOption {
+	return func(p *StreamPuller) {
+		p.partitionCount = count
+	}
+}
+
+// getStreamKey returns the partitioned stream name
+func (p *StreamPuller) getStreamKey(partition int) string {
+	return fmt.Sprintf("%s-%d{stream}", p.topicBase, partition)
+}
+
+// NewStreamPuller creates a new Redis Stream puller
+func NewStreamPuller(
+	client v3.Client,
+	subscription string, // Consumer group name
+	topic string, // Stream name
+	opts ...StreamPullerOption,
+) puller.Puller {
+	p := &StreamPuller{
+		redisClient:    client,
+		subscription:   subscription,
+		topicBase:      topic,
+		partitionCount: defaultStreamPartitionCount,
+		consumer:       fmt.Sprintf("%s-%s-%d", subscription, topic, time.Now().UnixNano()),
+		batchSize:      defaultBatchSize,
+		blockTime:      defaultBlockTime,
+		idleTime:       defaultIdleTime,
+		logger:         zap.NewNop(),
+		done:           make(chan struct{}),
+		messages:       make(chan *puller.Message),
+	}
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	p.logger = p.logger.Named("redis-stream-puller")
+
+	return p
+}
+
+// Pull reads messages from the stream and calls the handler for each message
+func (p *StreamPuller) Pull(ctx context.Context, handler func(context.Context, *puller.Message)) error {
+	p.mutex.Lock()
+	if p.closed {
+		p.mutex.Unlock()
+		return fmt.Errorf("redis stream puller is closed")
+	}
+
+	// Store the handler function for use in recoveryLoop
+	p.handler = handler
+	p.mutex.Unlock()
+
+	// Create consumer groups for all partitions
+	for partition := 0; partition < p.partitionCount; partition++ {
+		streamKey := p.getStreamKey(partition)
+
+		// Check if the consumer group exists before attempting to create it
+		groupExists, err := p.consumerGroupExists(ctx, streamKey, p.subscription)
+		if err != nil {
+			p.logger.Warn("Failed to check if consumer group exists",
+				zap.Error(err),
+				zap.String("subscription", p.subscription),
+				zap.String("stream", streamKey),
+			)
+		}
+
+		// Only create the group if it doesn't exist
+		if !groupExists {
+			err := p.redisClient.XGroupCreateMkStream(streamKey, p.subscription, "0")
+			if err != nil && err.Error() != "BUSYGROUP Consumer Group name already exists" {
```

Comment:
Maybe we should consider this way to avoid future issues in case Redis changes the error message.
```suggestion
			if err != nill && strings.Contains(strings.ToLower(err.Error()), "busygroup") {
```
or
```suggestion
			if err != nill && strings.Contains(strings.ToLower(err.Error()), "already exists") {
```

WDYT?